### ISSUE TITLE
Inheritance and Problem Facts

### ISSUE THREAD

**Issue by mmmvvvppp** [optapy/optapy#154]
I have a problem that is similar to the Job Shop problem wherein we have multiple types of tools that must be scheduled along with the location of the work, each of which are in limited supply.  The rest of the business code expects the tools have to have unique functions or attributes, thus I was trying to configure a domain whereby a base class Tool is defined and all specializations are derived from that.  The @planning_entity references the @problem_fact decorated Tool class, and specialized Tools are derived from that. 

I run into issues with this approach in post-8.23 releases of OptaPy whereby the @planning_id is not correctly resolved (even though a constraints verifier test validates the constraints and planning classes!).  This worked pre-8.23, albeit without the constraints verifier. 

Sample problem fact:
```python
@problem_fact # needed by the planning entity, or else a ValueError
class Tool(ABC):
    @abstractmethod
    def get_id(self) -> str:
        raise NotImplementedError("Calling function on abstract base class")

    @abstractmethod
    def __str__(self) -> str:
        raise NotImplementedError("Calling function on abstract base class")

@problem_fact
class Glue(Tool):
    def __init__(self, name: str, id: str, expiration_date) -> None:
        self.name = name
        self.id = str(id)
        self.expiration_date = expiration_date

    @planning_id  # If I mark the base class as well, I get a "The class has 2 members with a PlanningId annotation" error
    def get_id(self) -> str:
        return self.id

    def __str__(self) -> str:
        return f"Glue(id={self.id}, name={self.name})"
        
etc...
``````

Sample constraint:

```python
def constraint_tool_simultaneously(
    constraint_factory: ConstraintFactory,
):
    return (
        constraint_factory.for_each_unique_pair(
            ToolNeed,
            # ... if they support overlapping start/end times
            (Joiners.overlapping(lambda TN: TN.start_time, lambda RN: TN.end_time),),
        )
        # Then, if they share a selected tool, which can only be used by one job at once
        .filter(
            lambda TN1, TN2: TN1.selected_tool.get_id() == TN2.selected_tool.get_id()
        )
        # Then penalize it!
        .penalize("Overlapping jobs share a tool", HardSoftScore.ONE_HARD)
    )
```

I now get the following error:
`org.optaplanner.jpyinterpreter.types.errors.org.optaplanner.jpyinterpreter.types.errors.AttributeError: org.optaplanner.jpyinterpreter.types.errors.AttributeError: object 'Glue(id=100, name=Crazy)' does not have attribute 'get_id'`

If I try to take an alternative approach and move storage of the name and id attributes to the tool base class, and use a `super().__init__` call when initing the derived types, I get a `org.optaplanner.jpyinterpreter.types.errors.org.optaplanner.jpyinterpreter.types.errors.AttributeError: org.optaplanner.jpyinterpreter.types.errors.AttributeError: 'Tool' object has no attribute 'id'.` error.

**Comment by Christopher-Chianelli** [comment]
From the sounds of it, it looks like an issue with cloning (since constraint verifier works, and id is correctly resolved when tested in jpyinterpreter alone). Can I see your `@planning_entity` and `@planning_solution` classes?

**Comment by mmmvvvppp** [comment]
```python

@planning_entity
class ToolNeed:
    def __init__(
        self,
        tool_need_uuid: str,  # Unique ID for this instance
        list_of_suitable_tools: List[Tool],
        description: str,
        start_time: float = 0.0,
        end_time: float = 1000000000.0,
        selected_tool: Optional[Tool] = None,
    ):
        self.tool_need_uuid = tool_need_uuid
        self.description = description
        self.start_time = start_time
        self.end_time = end_time
        self.list_of_suitable_tools = list_of_suitable_tools
        self.selected_tool = selected_tool

    @planning_id
    def get_id(self) -> str:
        return self.tool_need_uuid

    @planning_variable(
        Tool, value_range_provider_refs=["list_of_suitable_tools"]
    )  # ,   nullable=True, )
    def get_selected_tool(self) -> Optional[Tool]:
        return self.selected_tool

    def set_selected_tool(self, new_selected_tool: Union[Tool, None]) -> None:
        self.selected_tool = new_selected_tool

    @problem_fact_collection_property(Tool)
    @value_range_provider(range_id="list_of_suitable_tools")
    def get_allowable_tools(self) -> List[Tool]:
        """Provide the solver with the problem space"""
        return self.list_of_suitable_tools

    def __str__(self) -> str:
        return (
            f"ToolNeed("
            f"id={self.tool_need_uuid}, "
            f"description={self.description}, "
            f"start_time={self.start_time}, "
            f"end_time={self.end_time}, "
            f"selected_tool={self.selected_tool})"
        )


@planning_solution
class PlanState:
    def __init__(
        self,
        tool_list: List[Tool],
        tool_requirements_list: List[ToolNeed],
        score: Optional[float] = None,
    ):
        self.tool_list = tool_list
        self.tool_requirements_list = tool_requirements_list
        self.score = score

    @problem_fact_collection_property(Tool)
    def get_tool_list(self) -> List[Tool]:
        return self.tool_list

    @planning_entity_collection_property(ToolNeed)
    def get_tool_requirements_list(self) -> List[ToolNeed]:
        return self.tool_requirements_list

    @planning_score(HardSoftScore)
    def get_score(self) -> HardSoftScore:
        return self.score

    def set_score(self, score: HardSoftScore) -> None:
        self.score = score

    def __str__(self) -> str:
        return (
            f"PlanState("
            # f"tool_list={format_list(self.tool_list)},\n"
            f"tool_requirements_list={format_list(self.tool_requirements_list)},\n"
            f"score={str(self.score)}"
            f")"
        )
```

I went ahead and uploaded a minimal example to https://github.com/mmmvvvppp/optapy155.  By default, the `requirements.txt` points to 8.23.0a0.

Using the [most recent commit](https://github.com/optapy/optapy/commit/11ce7594bd39eb4fcf5ff779605ab9cdb7e99416) to build/install a wheel, I was getting the following error:
`org.optaplanner.jpyinterpreter.types.errors.org.optaplanner.jpyinterpreter.types.errors.AttributeError: org.optaplanner.jpyinterpreter.types.errors.AttributeError: object 'Lock(id=10, name=Master)' does not have attribute 'get_id'`

**Comment by Christopher-Chianelli** [comment]
@mmmvvvppp I can confirm #158 fixes the issue on your given reproducer (I also fixed the ABC error, which seem to be caused by the incorrect type being passed down for subclasses of planning entity/problem facts) (additionally, it 32 times faster than 8.23 on my machine; probably even larger for a bigger dataset).

**Comment by mmmvvvppp** [comment]
@Christopher-Chianelli Do you have any insight into the cause of the slowdown?  I've noticed that as well and my suspicion is the `overlapping` call based on similar constraints.

**Comment by Christopher-Chianelli** [comment]
@mmmvvvppp 8.23 is the last version that does not have `jpyinterpreter`. Thus, upgrading to any version AFTER 8.23 will massively increase performance. As for the cause: `jpyinterpreter` prevents FFI calls, which eats up 99.99% of the runtime for small functions (and small functions are often used in optapy constraints). See https://github.com/optapy/optapy/pull/108 for full details.

**Comment by mmmvvvppp** [comment]
Ahh, my apologies, I misread your previous comment.  

Do you have any plans for a new release in the near future?

**Comment by Christopher-Chianelli** [comment]
Should be soon; we don't have CI yet for automating releases, so I need to push releases manually to PyPI (and hence why optapy is released at irregular intervals).

**Comment by Christopher-Chianelli** [comment]
Fixed by #158
